runtime/lib/capi/RuntimeCAPI.cpp

```diff
@@ -229,14 +229,22 @@ QirArray *__quantum__rt__qubit_allocate_array(int64_t num_qubits)
     RT_ASSERT(Catalyst::Runtime::CTX->getMemoryManager() != nullptr);
     RT_ASSERT(num_qubits >= 0);
 
-    QirArray *qubit_array = __quantum__rt__array_create_1d(sizeof(QubitIdType), num_qubits);
-    const auto &&qubit_vector =
+    // For first prototype, we just want to make this work.
+    // But ideally, I think the device should determine the representation.
+    // Essentially just forward this to the device library.
+    // And the device library can choose how to handle everything.
+    //
+    // Because QirArray* is opaque, we can return anything.
+    // But because this is a C function, we should return a C-like item.
+    // So, let's just return a pointer.
+    std::vector<QubitIdType> qubit_vector =
         Catalyst::Runtime::getQuantumDevicePtr()->AllocateQubits(num_qubits);
-    for (int64_t idx = 0; idx < num_qubits; idx++) {
-        *reinterpret_cast<QUBIT **>(__quantum__rt__array_get_element_ptr_1d(qubit_array, idx)) =
-            reinterpret_cast<QUBIT *>(qubit_vector[idx]);
-    }
-    return qubit_array;
+
+    // I don't like this copying.
```

Comment:
Hmm yeah that seems difficult to avoid because of the `AllocateQubits` function returning the vector on the stack 🤔 

We could always change the `QuantumDevice` interface of course, for example for measurement processes we pass in the allocated memory for the results. But then again these vectors will be fairly small so it shouldn't be too important.